### PR TITLE
Keyboard Driver

### PR DESCRIPTION
**Pull request by TheSheepKing**
Operational Keyboard Driver.
Minor upgrade of chaos-iso.sh.

### CODE REVIEW

File: drivers/keyboard.c [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+		}
+	}
+	apic_eoi();
+}
+
+/*
+** Sends the next char or loops until the user presses a key.
+*/
+char
+keyboard_next_input(void)
```
Comment (Arignir) [inline drivers/keyboard.c:130]: This code isn't protected against multi-threading. Locks are needed here to limit access to `input_buffer`, `input_write_idx` and `input_read_idx`.

File: drivers/keyboard.c [hunk 2]
```diff
@@ -0,0 +1,93 @@
+/* ------------------------------------------------------------------------ *\
+**
+**  This file is part of the Chaos Kernel, and is made available under
+**  the terms of the GNU General Public License version 2.
+**
+**  Copyright (C) 2017 - Benjamin Grange <benjamin.grange@epitech.eu>
+**
+\* ------------------------------------------------------------------------ */
+
+#include <kernel/init.h>
```
Comment (Arignir) [inline drivers/keyboard.c:10]: Ordering for includes is:

* chaosdef.h/chaoserr.h
* kernel/*.h
* arch/x86/*.h
* driver/*.h
* libc header

File: drivers/keyboard.c [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	[46] = 'c',
+	[47] = 'v',
+	[48] = 'b',
+	[49] = 'n',
+	[28] = '\n',
+	[14] = '\b',
+	[57] = ' ',
+};
+
+static volatile char input_buffer[PAGE_SIZE] = { 0 };
```
Comment (Arignir) [inline drivers/keyboard.c]: These variables don't need to be `volatile`.

File: drivers/keyboard.c [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+static volatile size_t input_write_idx = 0;
+static volatile size_t input_read_idx = 0;
+
+static void
+keyboard_ihandler(void)
+{
+	char code;
```
Comment (Arignir) [inline drivers/keyboard.c]: Shouldn't `code` be an unsigned char instead?

File: drivers/keyboard.c [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+	code = inb(KEYBOARD_IO_PORT);
+	if (!(code & 0x80)) { /* Only pressed keys */
+		code = fr_azerty_charset[(int)code];
+		if (code != '\0')
+		{
+			printf("%c", code);
+			input_buffer[input_write_idx] = code;
```
Comment (Arignir) [inline drivers/keyboard.c]: This is a bit dangerous as new input can erase old ones. Discarding new inputs would be better.

File: include/arch/x86/interrupts.h [hunk 6]
```diff
@@ -43,6 +43,9 @@ enum x86_interrupts
 	INT_IRQ0			= 0x20,
 
 	INT_PIT				= 0x20,
+
```
Comment (Arignir) [inline include/arch/x86/interrupts.h]: This blank line is irregular.

File: include/drivers/keyboard.h [hunk 7]
```diff
@@ -0,0 +1,12 @@
+#ifndef DRIVER_KEYBOARD_H_
+# define DRIVER_KEYBOARD_H_
+
+# include <chaosdef.h>
+
+# define INT_KEYBOARD			0x21
```
Comment (Arignir) [inline include/drivers/keyboard.h]: This is useless as it's already defined in `include/arch/x86/interrupts.h`.

File: include/drivers/keyboard.h [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+# define KEYBOARD_IO_PORT		0x60
```
Comment (Arignir) [inline include/drivers/keyboard.h]: This should be defined in `include/arch/x86/asm.h` instead.

File: include/drivers/keyboard.h [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+
+char
```
Comment (Arignir) [inline include/drivers/keyboard.h]: Prototypes doesn't have a new line after their return type.

**Review by Arignir — CHANGES_REQUESTED** [review]
In a general manner, this driver works. But ChaOS 2 is different from ChaOS 1, it's a bit more professionnal and well-written (even though it's not perfect).

Therefore, it would be great to set the _scan code set_ to avoid any surprises if the keyboard isn't in the _right scan code set_. Furethermore, supporting all characters and their uppercase/shiffted version would be even greater. Remeber to set the keyboard's LEDs if you do so.

Fix this and i'll merge.

File: drivers/keyboard.c (hunk 2 above)
Comment (TheSheepKing) [reply drivers/keyboard.c:10]: done.

File: drivers/keyboard.c (hunk 3 above)
Comment (TheSheepKing) [reply drivers/keyboard.c]: done.

File: drivers/keyboard.c (hunk 4 above)
Comment (TheSheepKing) [reply drivers/keyboard.c]: done.

File: drivers/keyboard.c (hunk 5 above)
Comment (TheSheepKing) [reply drivers/keyboard.c]: done.

File: include/arch/x86/interrupts.h (hunk 6 above)
Comment (TheSheepKing) [reply include/arch/x86/interrupts.h]: done.

File: include/drivers/keyboard.h (hunk 8 above)
Comment (TheSheepKing) [reply include/drivers/keyboard.h]: done.

File: include/drivers/keyboard.h (hunk 9 above)
Comment (TheSheepKing) [reply include/drivers/keyboard.h]: done.

File: drivers/keyboard.c [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+char
+keyboard_next_input(void)
+{
+	char c;
+
+	while (1)
+	{
+		mutex_acquire(&keyboard_lock);
+		if (input_read_idx == input_write_idx)
+			mutex_release(&keyboard_lock);
```
Comment (Arignir) [inline drivers/keyboard.c:138]: An interesting move here would be to reschedule if the condition failed, as it's very likely that the next input will not come in the next micro seconds.

The function to reschedule is `yield()`. You may want to read it's comment (to be sure not to use it incorrectly). It's in `kernel/thread.c`.

**Review by Arignir — CHANGES_REQUESTED** [review]
There is still the problem we discussed in `keyboard_ihandler`, with the keyboard buffer that you pull with `inb(KEYBOARD_IO_PORT)` may be non-empty when you leave `keyboard_ihandler`.

In addition, `mutex_acquire` may reschedule (see it's code in `include/kernel/mutex.h` and `include/kernel/semaphore.h`), with the current interruption still not marked as finished (`apic_eoi()` is placed after `mutex_acquire`).

Btw, please ping me when you want me to review your changes.

**Comment by sdicke** [comment]
In my opinion this implementation has some shortcomings:
1. It requires polling by another component.
2. Many PS/2 keyboard features are not covered.
3. It would be, probably, better to implement generic PS/2 port features and use them in the keyboard driver to allow code reuse when implement drivers for another PS/2 devices like mice.

**Comment by Arignir** [comment]
Hello @sdicke ,

Thank your for your interest in ChaOS! You're the very first person external to the project that reaches to us, and we value your feedback.

This is a school project and we are starting to get short on deadlines. Therefore, we have to take a decision between making something clean and maintainable, and something a bit dirty than can provide the mandatory features that we need. This is the trade-off we are facing.

This keyboard driver gives a solution to one of the most important feature missing righ-now in ChaOS: User interaction. We can't deliver our project without that. 

Thus, we decided to have a quick&dirty keyboard driver until the project is delivered, and i'll improve it later when i'll have the time & energy.

**Pull request merged**